crates/ide_completion/src/render.rs

```diff
@@ -979,12 +990,54 @@ fn main() {
                         detail: "S",
                         relevance: CompletionRelevance {
                             exact_name_match: true,
-                            exact_type_match: false,
+                            exact_type_match: true,
                         },
                         ref_match: "&mut ",
                     },
                 ]
             "#]],
         )
     }
+
+    #[test]
+    fn suggest_deref() {
+        check_relevance(
+            r#"
+#[lang = "deref"]
+trait Deref {
+    type Target;
+    fn deref(&self) -> &Self::Target;
+}
+
+struct S;
+struct T(S);
+
+impl Deref for T {
+    type Target = S;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+fn foo(s: &S) {}
+
+fn main() {
+    let t = T(S);
+    let m = 123;
+
+    foo($0);
```

Comment:
Isn't `t` not a type match, as in `foo(t)` would not compile? I think ideally we'd suggest `&t` and mark that as an exact type match.

Reply:
If I understand correctly, we check if we need to put reference as a param in https://github.com/rust-analyzer/rust-analyzer/blob/9674490b45a931399a9a05cd3a4c7e8fae9ec2a8/crates/ide_completion/src/render.rs#L257-L269
To make things like that possible, I might need to pass some kind of flag from `compute_relevance`, that the type matches because of deref, and that `ref_match` needs to be executed on builder

Reply:
Rather than passing a flag around I think you *might* (sorry I'm still coming up to speed on this) be able to make your change directly in the code you referenced above. As in `t` shouldn't itself be an exact type match, but `t` should have a `ref_match`. `ref_match` are automatically tagged as `exact_type_match`.

Reply:
I'm not sure if I'm doing something incorrectly here, or just found an error. `ref_match` does trigger `exact_type_match`, howerer, I am not able to find any call site for it (RA gives no results for reference search). By manually replacing `it.relevance()` with `it.ref_match().map(|m| m.1)` in tests I'm giving myself a successful test run

Reply:
After changes proposed above it works fine, suggesting the right reference type when needed. But I'm still unsure about what to do with tests? Push with replacements stated above?

Reply:
https://github.com/rust-analyzer/rust-analyzer/blob/7accf6bc37c059a83a58c82f463f02a02ed2226f/crates/rust-analyzer/src/to_proto.rs#L241-L254

This is the code that calls `ref_match()`. Although it does look like the test code in `crates/ide_completion/src/render.rs` doesn't look at the `ref_match`, and I think it should. 

Requiring `to_proto` to deal with this ref_match stuff feels to me like `ide_completions::completions` isn't quite providing the right public API, but that is a discussion that should be had outside this PR probably.

Reply:
Oh, so that explains why it worked when launching test environment. I guess replacing `it.relevance()` with `it.ref_match().map(|m| m.1).unwrap_or(it.relevance())` should be fine then (if we ignore references in completion labels in tests)

Reply:
> But I'm still unsure about what to do with tests.

Short of updating the public API for completions as described above, which I think is out of scope for this PR, I think I'd be in favor of updating `check_relevance` to `flat_map` the completions to add in the `ref_match`, as is done in `to_proto`. That way the tests can see both the original match and the ref match (in cases where it exists), which seems appropriate to me.

Reply:
Actually nvm, it's better to check references in tests too, so I'll rewrite them

Reply:
Got it